### ISSUE TITLE
light.exe returns -2146232576 in Windows 2016 Docker container

### ISSUE THREAD

**Issue by eerhardt** [wixtoolset/issues#5213]
I am trying to get our continuous integration system working inside a Windows Docker container.

I set up a Windows 2016 container host following these [instructions](https://msdn.microsoft.com/en-us/virtualization/windowscontainers/quick_start/azure_setup).

I tried running the [dotnet/cli](https://github.com/dotnet/cli) build inside of the docker container using the following dockerfile:

```
#
# Copyright (c) .NET Foundation and contributors. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for full license information.

# Dockerfile that creates a container suitable to build dotnet-cli on Windows

FROM windowsservercore

# Install Git
RUN powershell.exe invoke-webrequest https://github.com/git-for-windows/git/releases/download/v2.7.1.windows.1/Git-2.7.1-64-bit.exe -Outfile c:\git.exe
RUN C:\git.exe /SILENT /COMPONENTS="icons,ext\reg\shellhere,assoc,assoc_sh"
RUN DEL C:\git.exe

# Install CMake
RUN powershell.exe invoke-webrequest https://cmake.org/files/v3.4/cmake-3.4.3-win32-x86.exe -Outfile c:\cmake.exe
RUN C:\cmake.exe /S
RUN DEL C:\cmake.exe

# Install C++ VS 2015 Community Update 1
RUN powershell.exe invoke-webrequest http://download.microsoft.com/download/5/7/A/57A99666-126E-42FA-8E70-862EDBADD215/vs_community.exe -Outfile c:\vs.exe
RUN powershell.exe Start-Process C:\vs.exe -ArgumentList ('/q', '/norestart', '/InstallSelectableItems', 'NativeLanguageSupport_VC') -Wait
RUN DEL C:\vs.exe

RUN setx /M PATH "%PATH%;C:\Program Files\Git\cmd;C:\Program Files (x86)\CMake\bin"

# Set working directory
WORKDIR D:/
```

And when running the docker container, using `-v "C:\CLIRepoRoot:D:"` to map the git repo from the Host VM on the container.

Using both `3.10.0.2103-pre1` and `3.10.1` versions of the WiX toolset, the following line failed when running in the container:

```PS1
    .\light.exe -nologo `
        -cultures:en-us `
        bundle.wixobj `
        -ext WixBalExtension.dll `
        -ext WixUtilExtension.dll `
        -ext WixTagExtension.dll `
        -b "$AuthWsxRoot" `
        -out $DotnetBundleOutput | Out-Host
```

with exit code -2146232576.

I tried running this on the Windows 2016 host machine, and it passed successfully. So it has to be something on the docker container that doesn't work.

The powershell script that runs the WiX tools is here: https://github.com/dotnet/cli/blob/rel/1.0.0/packaging/windows/generatemsi.ps1#L162

Note that earlier in the script we ran heat.exe and candle.exe to success. We were even able to run light.exe on our .msi, which succeeded. However, only when light.exe was run for our bundle it failed with this exit code. `light.exe /?` is able to be run on the container and it prints out the help correctly.

**Comment by barnson** [comment]
Note that there are no official WiX v3.x NuGet packages. You should avoid them in general. (We just heard today of a user who relied on an unofficial NuGet package that was an incorrectly-rebuilt fork of WiX.) You can use the same download-and-install approach to install official, signed WiX binaries.

That said...Can you capture a stack trace? The official builds have .pdbs available in a .zip file you can copy in place and get good traces.

**Comment by barnson** [comment]
BTW, it's important that you use WiX v3.10.2 because it [mitigates DLL hijacking vulnerabilities in Windows](https://www.firegiant.com/blog/2016/1/20/wix-v3.10.2-released/).

**Comment by eerhardt** [comment]
Where can I find the "official" build? If there are no NuGet packages, what is the recommended way to consume the WiX tools from a git repo (i.e. I can't check in binary files)? Should I be downloading the .zip file from https://wix.codeplex.com/downloads/get/1540241? I took that link from https://wix.codeplex.com/releases/view/619491.

Capturing a stack trace is going to be a little hard, because you only get to run 1 command in a Docker container. And you only can use the command line. You don't get any GUI applications - no VS or WinDbg GUI. So I will need to first install WinDbg on the container, and use WinDbg on the command line.

Is there some sort of logging or verbosity switch in light.exe that I can try? Dump a log to a file or to the standard out?

**Comment by barnson** [comment]
Yes, that's the binaries .zip for v3.10.2. If you grab https://wix.codeplex.com/downloads/get/1540242 as well and unzip it so the .pdb files are siblings of the WiX .exes and .dlls, a crash will give a stack trace with line numbers. No GUIs or Winbag required. (Note that I'm guessing it's a crash -- that error code is an undocumented CLR initialization failure.)

**Comment by barnson** [comment]
And, of course, by "crash," I mean unhandled exception. Same thing. :)

**Comment by eerhardt** [comment]
I put v3.10.2 along with the .pdbs on the machine and now I am getting a failure in the first run of light.exe to create the .msi.  Before I was getting the error on the 2nd run when creating the bundle .exe file.

```
light.exe : error LGHT0001 : Access is denied. (Exception from HRESULT: 0x80070005 (E_ACCESSDENIED))

Exception Type: System.UnauthorizedAccessException

Stack Trace:
   at Microsoft.Tools.WindowsInstallerXml.Cab.Interop.NativeMethods.EnumerateCab(String cabinet, PFNNOTIFY notify)
   at Microsoft.Tools.WindowsInstallerXml.BinderFileManager.ResolveCabinet(FileRowCollection fileRows, String& cabinetPath) in C:\build\work\eca3d12b\wix3\src\tools\wix\BinderFileManager.cs:line 511
   at Microsoft.Tools.WindowsInstallerXml.Binder.CreateCabinetWorkItem(Output output, String cabinetDir, MediaRow mediaRow, FileRowCollection fileRows, ArrayList fileTransfers) in C:\build\work\eca3d12b\wix3\src\tools\wix\Binder.cs:line 7167
   at Microsoft.Tools.WindowsInstallerXml.Binder.CreateCabinetFiles(Output output, FileRowCollection fileRows, ArrayList fileTransfers, MediaRowCollection mediaRows, String layoutDirectory, Boolean compressed, AutoMediaAssigner autoMediaAssigner) in C:\build\work\eca3d12b\wix3\src\tools\wix\Binder.cs:line 7082
   at Microsoft.Tools.WindowsInstallerXml.Binder.BindDatabase(Output output, String databaseFile) in C:\build\work\eca3d12b\wix3\src\tools\wix\Binder.cs:line 1968
   at Microsoft.Tools.WindowsInstallerXml.Binder.Bind(Output output, String file) in C:\build\work\eca3d12b\wix3\src\tools\wix\Binder.cs:line 658
   at Microsoft.Tools.WindowsInstallerXml.Tools.Light.Run(String[] args) in C:\build\work\eca3d12b\wix3\src\tools\light\light.cs:line 462
Light failed with exit code 1.
```

**Comment by eerhardt** [comment]
I was able to limp past the above error by using the v3.10.1 light.exe from the NuGet packages that we were previously using to build the .msi. And then using v3.10.2 light.exe that has the pdbs next to it. I still got the same result without getting anything logged to the console:

```
C:\wix>.\light.exe -nologo -cultures:en-us bundle.wixobj -ext WixBalExtension.dll -ext WixUtilExtension.dll -ext WixTagExtension.dll -b "D:\packaging\windows" -out D:\artifacts\win10-x64\packages\dnvm\dotnet-win-x64.exe

C:\wix>echo %errorlevel%
-2146232576
```

**Comment by barnson** [comment]
At this point on a "normal" machine, I'd try a dump or a ProcMon log. Not sure how the container influences CLR loading.

**Comment by eerhardt** [comment]
Well, I found one way around this "failed to load runtime" error that I'm seeing, and that is to run light.exe just before the real call:

```
#I added this line
    .\light.exe /? | Out-Null

    .\light.exe `
        -cultures:en-us `
        bundle.wixobj `
        -ext WixBalExtension.dll `
        -ext WixUtilExtension.dll `
        -ext WixTagExtension.dll `
        -b "$AuthWsxRoot" `
        -out $DotnetBundleOutput | Out-Host
```

Adding the dummy call to `light.exe /?` allows the second light.exe call to load. There must be an issue here with Windows Docker containers. I'll follow up with the Windows and .NET teams.

However, the bundle still isn't getting created with the second call. I am now getting an UnauthorizedAccessException:

```
Running light for bundle..
Windows Installer XML Toolset Linker version 3.10.1.2213
Copyright (c) Outercurve Foundation. All rights reserved.

light.exe : error LGHT0001 : Access to the path 'D:\artifacts\win10-x64\packages\dnvm\dotnet-win-x64.1.0.0.1414.msi' is denied.

Exception Type: System.UnauthorizedAccessException

Stack Trace:
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, String msgPath, Boolean bFromProxy)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, Boolean useAsync)
   at System.IO.FileInfo.OpenRead()
   at Microsoft.Tools.WindowsInstallerXml.Common.GetFileHash(FileInfo fileInfo)
   at Microsoft.Tools.WindowsInstallerXml.PayloadInfoRow.ResolvePayloadInfo(IPayloadInfo payloadInfo)
   at Microsoft.Tools.WindowsInstallerXml.PayloadInfoRow.FillFromPayloadRow(Output output, Row payloadRow)
   at Microsoft.Tools.WindowsInstallerXml.Binder.BindBundle(Output bundle, String bundleFile)
   at Microsoft.Tools.WindowsInstallerXml.Binder.Bind(Output output, String file)
   at Microsoft.Tools.WindowsInstallerXml.Tools.Light.Run(String[] args)
Light failed with exit code 1.
```

@barnson - any thoughts? Maybe generating the .msi isn't setting the file up correctly?

When I try running the .msi `msiexec /a artifacts\win10-x64\packages\dnvm\dotnet-win-x64.1.0.0.1414.msi` I get an error:

```
[Window Title]
Windows Installer

[Main Instruction]
This installation package could not be opened.  Verify that the package exists and that you can access it, or contact the application vendor to verify that this is a valid Windows Installer package.

[OK]
```

I am running as admin.

So somehow the file permissions/attributes on the .msi must not be correct.

**Comment by barnson** [comment]
The -sacl Light.exe suppresses the default behavior of resetting ACLs. Maybe its approach is wrong in a container?

**Comment by eerhardt** [comment]
Some more info after further investigation:

I am starting to believe that it is the _previous_ call to light.exe to create the .msi that has the problem. Evidence for this:

1. The .msi file cannot be accessed inside nor outside of the docker container. I always get "access is denied" when trying to work with that file.
1. The very next call to "light.exe" fails with -2146232576 "failed to load the runtime", but if you turn around and call "light.exe" immediately after, it loads just fine.
1. It isn't just the next call to "light.exe", it appears to be _any_ managed .exe can't be loaded next. I injected a call to `heat.exe /?` in between light.exe for the .msi and light.exe for the bundle, and I got `heat.exe failed with exit code -2146232576.`

**Comment by eerhardt** [comment]
I also tried running `ConsoleApplication18.exe` (which is just a console app that Console.WriteLine(Environment.Version);), and that also fails with exit code -2146232576. So now I firmly believe it is the call to light.exe that generates the .msi that is causing the machine to get into a state where it can't load the next managed executable. But then once it fails for one managed .exe, the state somehow gets cleared and the next managed .exe can be loaded successfully.

I've pinged the Windows Container and .NET Runtime team internally for this.

**Comment by barnson** [comment]
Light.exe uses a lot of P/Invoke and reg-free ComImport so it's not exactly a typical managed executable. That could contribute to the problem.

**Comment by eerhardt** [comment]
I'm fairly confident this is a Windows Containers issue with the registry. I have been working with the Windows team on this.
It appears at this moment, any managed executable will raise this error because it cannot find the .NET runtime in the registry.

Closing this issue since it doesn't appear to be a WiX issue.

**Comment by timotei** [comment]
@eerhardt Hi. Have you managed to find the issue? I had the same issue, which I fixed by installing the .NET 3.5 feature (directly via the cab from the installer)

**Comment by eerhardt** [comment]
> Have you managed to find the issue?

We traced it down to a registry issue with Windows 2016 TP4.  At this exact point in my build, no matter what .NET executable was trying to be executed (light.exe or otherwise), the .exe would fail with this error. And if you just ignored the error and tried the .NET .exe again, it would succeed.

I haven't tried it on TP5 yet to ensure the registry issue was fixed.
